src/rdkafka_telemetry.c

```diff
@@ -0,0 +1,785 @@
+                    rk->rk_telemetry.accepted_compression_types[i];
+
+                switch (compression_type) {
+#if WITH_ZLIB
+                case RD_KAFKA_COMPRESSION_GZIP:
+                        r = rd_kafka_compress_gzip(rkb, payload, payload_len,
+                                                   compressed_payload,
+                                                   compressed_payload_size);
+                        compression_used = RD_KAFKA_COMPRESSION_GZIP;
+                        break;
+#endif
+                case RD_KAFKA_COMPRESSION_LZ4:
+                        // TODO: Using 0 for compression level for now.
+                        r = rd_kafka_lz4_compress_direct(
+                            rkb, 0, payload, payload_len, compressed_payload,
+                            compressed_payload_size);
+                        compression_used = RD_KAFKA_COMPRESSION_LZ4;
+                        break;
+#if WITH_ZSTD
+                case RD_KAFKA_COMPRESSION_ZSTD:
+                        // TODO: Using 0 for compression level for now.
+                        r = rd_kafka_zstd_compress_direct(
+                            rkb, 0, payload, payload_len, compressed_payload,
+                            compressed_payload_size);
+                        compression_used = RD_KAFKA_COMPRESSION_ZSTD;
+                        break;
+#endif
+#if WITH_SNAPPY
+                case RD_KAFKA_COMPRESSION_SNAPPY:
+                        r = rd_kafka_compress_snappy(rkb, payload, payload_len,
+                                                     compressed_payload,
+                                                     compressed_payload_size);
+                        compression_used = RD_KAFKA_COMPRESSION_SNAPPY;
+                        break;
+#endif
+                default:
+                        break;
+                }
+                if (compression_used != RD_KAFKA_COMPRESSION_NONE && r == 0) {
+                        rd_kafka_dbg(
+                            rk, TELEMETRY, "PUSH",
+                            "Compressed payload of size %" PRIusz " to %" PRIusz
+                            " using compression type "
+                            "%s",
+                            payload_len, *compressed_payload_size,
+                            rd_kafka_compression2str(compression_used));
+                        rd_free(payload);
+                        return compression_used;
+                }
+        }
+        if (compression_used != RD_KAFKA_COMPRESSION_NONE && r == -1) {
+                rd_kafka_dbg(rk, TELEMETRY, "PUSHERR",
+                             "Failed to compress payload with available "
+                             "compression types");
+        }
+        rd_kafka_dbg(rk, TELEMETRY, "PUSH", "Sending uncompressed payload");
+
+        *compressed_payload      = payload;
+        *compressed_payload_size = payload_len;
+        return RD_KAFKA_COMPRESSION_NONE;
+}
```

Comment:
You can use existing compression functions here or set the pointers to `payload->rbuf_wpos->seg_p` and  `payload->rbuf_wpos->seg_of` in case no compression was applied.
```suggestion
static rd_kafka_compression_t
rd_kafka_push_telemetry_payload_compress(rd_kafka_t *rk,
                                         rd_kafka_broker_t *rkb,
                                         rd_buf_t *payload,
                                         void **compressed_payload,
                                         size_t *compressed_payload_size) {
        rd_kafka_compression_t compression_used = RD_KAFKA_COMPRESSION_NONE;
        rd_slice_t payload_slice;
        size_t i;
        rd_kafka_resp_err_t r = RD_KAFKA_RESP_ERR_NO_ERROR;
        rd_slice_init_full(&payload_slice, payload);

        for (i = 0; i < rk->rk_telemetry.accepted_compression_types_cnt; i++) {
                rd_kafka_compression_t compression_type =
                    rk->rk_telemetry.accepted_compression_types[i];

                switch (compression_type) {
#if WITH_ZLIB
                case RD_KAFKA_COMPRESSION_GZIP:
                        /* TODO: Using 0 for compression level for now. */
                        r = rd_kafka_gzip_compress(rkb, 0, &payload_slice,
                                                   compressed_payload,
                                                   compressed_payload_size);
                        compression_used = RD_KAFKA_COMPRESSION_GZIP;
                        break;
#endif
                case RD_KAFKA_COMPRESSION_LZ4:
                        /* TODO: Using 0 for compression level for now. */
                        r = rd_kafka_lz4_compress(
                            rkb, rd_true, 0, &payload_slice, compressed_payload,
                            compressed_payload_size);
                        compression_used = RD_KAFKA_COMPRESSION_LZ4;
                        break;
#if WITH_ZSTD
                case RD_KAFKA_COMPRESSION_ZSTD:
                        /* TODO: Using 0 for compression level for now. */
                        r = rd_kafka_zstd_compress(rkb, 0, &payload_slice,
                                                   compressed_payload,
                                                   compressed_payload_size);
                        compression_used = RD_KAFKA_COMPRESSION_ZSTD;
                        break;
#endif
#if WITH_SNAPPY
                case RD_KAFKA_COMPRESSION_SNAPPY:
                        r = rd_kafka_snappy_compress(rkb, &payload_slice,
                                                     compressed_payload,
                                                     compressed_payload_size);
                        compression_used = RD_KAFKA_COMPRESSION_SNAPPY;
                        break;
#endif
                default:
                        break;
                }
                if (compression_used != RD_KAFKA_COMPRESSION_NONE &&
                    r == RD_KAFKA_RESP_ERR_NO_ERROR) {
                        rd_kafka_dbg(
                            rk, TELEMETRY, "PUSH",
                            "Compressed payload of size %" PRIusz " to %" PRIusz
                            " using compression type "
                            "%s",
                            payload->rbuf_size, *compressed_payload_size,
                            rd_kafka_compression2str(compression_used));
                        return compression_used;
                }
        }
        if (compression_used != RD_KAFKA_COMPRESSION_NONE &&
            r != RD_KAFKA_RESP_ERR_NO_ERROR) {
                rd_kafka_dbg(rk, TELEMETRY, "PUSH",
                             "Failed to compress payload with available "
                             "compression types");
        }
        rd_kafka_dbg(rk, TELEMETRY, "PUSH", "Sending uncompressed payload");

        *compressed_payload      = payload->rbuf_wpos->seg_p;
        *compressed_payload_size = payload->rbuf_wpos->seg_of;
        return RD_KAFKA_COMPRESSION_NONE;
}
```

Reply:
Updated